File: generated_code/GeminiPro/Python/035_SplayTree.py
import sys

class SplayTree:
    """
    Implements a Splay Tree, a self-balancing binary search tree.

    This class provides a dictionary-like set for storing unique integers.
    The key feature is the splaying operation, which moves a recently accessed
    element to the root of the tree, optimizing for future accesses of the
    same element. The tree does not store duplicate keys.
    """

    class _Node:
        """A private class representing a node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _splay(self, key):
        """
        Performs the splaying operation on the given key.

        This method rearranges the tree so that the node with the given key,
        or the node that would be its parent if the key is not present, is
        moved to the root of the tree. This is a top-down splaying implementation.
        """
        if not self.root:
            return

        # Use a dummy node to simplify linking of left and right subtrees
        header = self._Node(None)
        left_tree_max, right_tree_min = header, header
        current = self.root

        while True:
            if key < current.key:
                if not current.left:
                    break
                # Zig-Zig case (left-left): Rotate right
                if key < current.left.key:
                    y = current.left
                    current.left = y.right
                    y.right = current
                    current = y
                    if not current.left:
                        break
                # Link current node to the right tree of the new assembly
                right_tree_min.left = current
                right_tree_min = current
                current = current.left
            elif key > current.key:
                if not current.right:
                    break
                # Zig-Zig case (right-right): Rotate left
                if key > current.right.key:
                    y = current.right
                    current.right = y.left
                    y.left = current
                    current = y
                    if not current.right:
                        break
                # Link current node to the left tree of the new assembly
                left_tree_max.right = current
                left_tree_max = current
                current = current.right
            else:  # key is found
                break

        # Reassemble the tree
        left_tree_max.right = current.left
        right_tree_min.left = current.right
        current.left = header.right
        current.right = header.left
        self.root = current

    def search(self, key):
        """
        Searches for a key in the tree and performs the splaying operation.

        After this operation, if the key is found, the node containing it
        becomes the new root. If not found, the last accessed node (the
        would-be parent) becomes the new root.

        Args:
            key (int): The key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False
        
        self._splay(key)
        
        return self.root.key == key

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the node is splayed to the root.
        If the key does not exist, it is inserted and becomes the new root.

        Args:
            key (int): The key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the tree to bring the closest node to the root
        self._splay(key)

        # If key is already at the root, we're done
        if self.root.key == key:
            return

        # Otherwise, insert the new node at the root and split the tree
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None
        
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree.

        If the key is found, it is deleted. The tree is rebalanced by
        joining the two resulting subtrees. If the key is not found,
        the tree is splayed based on the search for the key, but no
        node is removed.

        Args:
            key (int): The key to delete.
        """
        if not self.root:
            return

        # Splay the key to the root
        self._splay(key)

        # If key is not at the root, it's not in the tree, so do nothing
        if self.root.key != key:
            return

        # The node to delete is now the root.
        # We join its left and right subtrees.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left child, the right subtree becomes the new tree
            self.root = right_subtree
        else:
            # Make the left subtree the main tree
            self.root = left_subtree
            # Splay for the largest element in this new tree.
            # Since `key` is larger than any element in the left subtree,
            # splaying for it will bring the max element to the root.
            self._splay(key)
            # Attach the original right subtree
            self.root.right = right_subtree

File: generated_code/GeminiPro/Java/035_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * Splay trees are self-balancing binary search trees with the additional property that
 * recently accessed elements are quick to access again. The splay operation moves an
 * accessed element to the root of the tree.
 */
public class SplayTree {

    /**
     * Represents a node in the splay tree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructs a node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * Performs the top-down splay operation. It brings the node with the given key
     * (or the last accessed node on the search path if the key is not found)
     * to the root of the tree.
     *
     * @param node The root of the tree/subtree to splay.
     * @param key The key to splay for.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        if (node == null) {
            return null;
        }

        // Dummy node to act as a header for the left and right subtrees being built.
        // The key in the dummy node is irrelevant.
        Node dummy = new Node(0);
        Node leftTree = dummy;
        Node rightTree = dummy;

        while (true) {
            if (key < node.key) {
                if (node.left == null) {
                    break;
                }
                // Zig-Zig case (right rotation)
                if (key < node.left.key) {
                    node = rightRotate(node);
                    if (node.left == null) {
                        break;
                    }
                }
                // Link current root to the right tree
                rightTree.left = node;
                rightTree = node;
                node = node.left;
            } else if (key > node.key) {
                if (node.right == null) {
                    break;
                }
                // Zig-Zig case (left rotation)
                if (key > node.right.key) {
                    node = leftRotate(node);
                    if (node.right == null) {
                        break;
                    }
                }
                // Link current root to the left tree
                leftTree.right = node;
                leftTree = node;
                node = node.right;
            } else {
                // Key found
                break;
            }
        }

        // Reassemble the tree
        leftTree.right = node.left;
        rightTree.left = node.right;
        node.left = dummy.right;
        node.right = dummy.left;

        return node;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree remains unchanged but the existing node is splayed.
     * After insertion of a new key, the new node becomes the root of the tree.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. The closest node becomes the root.
        root = splay(root, key);

        // If key is already present, do nothing more.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Searches for a key in the splay tree.
     * This operation performs splaying. If the key is found, it is splayed to the root.
     * If not found, the last accessed node on the search path is splayed to the root.
     *
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not in the tree, the tree is splayed around the key's would-be position
     * but remains structurally unchanged regarding its elements.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the key to the root.
        root = splay(root, key);

        // If the key is not at the root after splaying, it wasn't in the tree.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Find the maximum element in the left subtree by splaying for a large value.
            // This makes the maximum element the new root of the left subtree.
            leftSubtree = splay(leftSubtree, Integer.MAX_VALUE);
            
            // The new root of the left subtree (the max element) now has no right child.
            // Attach the original right subtree to it.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}